.travis.yml

```diff
@@ -1,10 +1,20 @@
 sudo: required
-services:
-  - docker
 
-before_install:
-- docker pull edbrown23/sysadmin:1
-- docker build . -t sysadmin_tester
+env:
+  global:
+  - secure: Nk0lL86bdT9HlUzgFGz3XCckTdss0cjZY9mzbNh2XttyMxRws4ar62PNuYHHgM8h4NIuIzdiCC+ulrTVQgqx4ui/2SxkAUgFrSKlYrdfKEIYgTEa/RkRCYrLs40OEqOgxKDhwudi6c3Vx5J3dWBllrFKIU4C3heOJIibVmvQrarZojQc5bAimKFAXl9ZobezD3VtkW+az9lSzYN93ypdo3yAflTPPWfvF/rDVTz2avJsExtPv43f+J1mQj+AiiU4L0qXkHwBV+a+A/ZqHjmreSyUmShFXX7bn9qcdb6SKQBbS/i/qGahUVrbJ4lPkpjAHEmZ8fi6yz8nj3GoahcA9sQ6NgpLtSPhy47GLeDwdMoOTq/lrwKPPAwPTUjrGeM/VfKBFiWa12J30EMH5Nv2EzpNlY3bEodRIQSiN02G6xzfyK7XDyg4nUMraIPwCs3lAv4wEl3LIImMkyOnQexE0oFkCEAOMHde0VszTUphPnNDLMY6VaMls1JVzTTgWO8e1GGQVB4lISPDLehTMLnJZE43zrHrhEf7oA3O6YtRezpuB+7wJ3YfNsPQh5FQ7GMAhAufH5ss/Ql05D8WfCjUmaFkbagixcUozaNKfct6MFgWRUw5fdXAd37A6DoCeRXoO9vH7nTDm443aG2uZYavdGGjNCPTnq2atOjHj6itRbc=
+  - secure: CrKt1gN2mcR8quetGs6CWSVSNPQoG0irsGztYhkGZusjC9blZzPyLndNwfyVDqt3m0dSQYow9p++Z1au08EhB29yYPb93j6KDCVNK+A0wWKY1gExA9En6sp1k9WyzHFSCs/Xq42vT4tNjTZinw+Qk1Wwc/bqeL1zpbQ8KmjsMPrdGqJZiuRlH9KNAHUuAFXR+yffokEMb43yD+wGfKfN0DAe4Z9O3KgWsoCg8FLiPDDjuf56/RjQH6vc2Jfn/8TjNY1h7ZIiWMwLrxqvKedt9bSRiafApgSCR+IWJzmo9ZsDHyV0R+wxKhZRHTF4kfAejDOKopMeOqaybbw/HywpOVc8z7AHMfRxKMt3Oiz6CP3CDq/VZoK9uQvn6m/+tnQ1XvEWwh3Z+g0GnKFJhMB03AlBJu1ngv5wPLzzfX81Mt02DLMOhlzeKWBRvohgovF74u3mGCn6uS+NG0bw8wqKImLuqqeQCJjSHNUXASZwLFv95LURhZu3GkRquupqiJ4cR2SESukPiBdhI4imkSCYuwGpEuv/4NJuBqM5otUHygf6XhUSgnLqqfAOoDdPYRYXt2VPvgbvr2pJ6hMbdTGbsJdFHW4nCU5abQ0yvxqrw8NrgfHYwkvYTeTzr1oDc+P0bggw1TrQUmoFgBtmie8iB0NI8XeOgot5ShlvEfjyuMo=
 
+services:
+- docker
+before_install:
+- docker pull starryoss/sysadmin:2 || docker build -t starryoss/sysadmin:2 docker_builds/
+- docker build --build-arg=USERID=`id -u` . -t sysadmin_tester
 script:
-- docker run sysadmin_tester /bin/bash -c "cd /var/tmp/workdir/sysadmin; rm -rf build; mkdir -p build; cd build; cmake ..; make check"
+- docker run -it --rm -v `pwd`:/home/user/sysadmin -u user --workdir /home/user/sysadmin sysadmin_tester /bin/bash -c "rm -rf /home/user/sysadmin/build && mkdir -p /home/user/sysadmin/build && cd /home/user/sysadmin/build && cmake .. && make check && make "
+after_success:
+- if [ "$TRAVIS_BRANCH" == "master" ]; then
+  docker login -u="$DOCKER_USERNAME" -p "$DOCKER_PASSWORD" ;
+  docker push starryoss/sysadmin:2 ;
```

Comment:
should the version be based on a tag?

Reply:
Also semver'd?

Reply:
This is a bit hacky but It should work for now. When we say push it will usually just see that it already has the image and exit. 

I also want builds to be fast and if we use `git tag` I feel like we will be rebuilding on each commit. Does that make sense?

Reply:
git tag wouldn't rebuild on every commit, it would rebuild on every release. However - we probably _do_ want this to be rebuilt on every commit, that is usually what a `latest` docker tag points to.

The way I'd imagine this working:
* If a PR changes the dockerfile, we want to ensure the build is tested and that we run the tests in that container. There is probably some caching we can do to optimize the build. Pulling down `latest` should allow most builds to avoid rebuilding the image. Ultimately, all builds should build the image.
* On master commits, we push a `starryoss/sysadmin:latest` image
* On tagged commits, we push a `starryoss/sysadmin:<tag>` image